File: AdventOfCode2021/d23_typescript/d23.ts
import * as readline from "readline";

type elt = 'A'|'B'|'C'|'D';
type hallElt = '.'|elt;

function hallIdx(stck: number): number {
    return 2 * (1 + stck);
}

function isRoomDoor(idx: number): boolean {
    return 0 < idx && idx < 10 && idx % 2 == 0
}

function openHallSlots(hall: hallElt[], idx: number): number[] {
    const s: number[] = [];
    let w = idx;
    while(w > 0 && hall[w-1] == '.') {
        w--;
        s.push(w);
    }
    w = idx;
    while(w < 10 && hall[w+1] == '.') {
        w++;
        s.push(w);
    }
    return s;
}

function movesToHall(hall: hallElt[], idx: number): number[] {
    return openHallSlots(hall, idx).filter(x => !isRoomDoor(x));
}

function stackPseudoSolved(stcks: elt[][], idx: number): boolean {
    const tgt: elt = String.fromCharCode("A".charCodeAt(0) + idx) as elt;
    for(const st of stcks[idx])
        if(st !== tgt)
            return false;
    return true;
}

function stackSolved(stcks: elt[][], idx: number, stackh: number): boolean {
    return stcks[idx].length == stackh && stackPseudoSolved(stcks, idx);
}

function solved(stcks: elt[][], stackh: number): boolean {
    for(let idx = 0; idx < stcks.length; idx++)
        if(!stackSolved(stcks, idx, stackh))
            return false;
    return true;
}

function toKey(stcks: elt[][], hall: hallElt[]): string {
    return `([${stcks.map(stck=>`[${stck}]`)}], "${hall}")`;
}

const costMp = {
    'A': 1,
    'B': 10,
    'C': 100,
    'D': 1000,
};

const memo: Map<string, number> = new Map();
function solve(stcks: elt[][], hall: hallElt[], stackh: number): number {
    if(solved(stcks, stackh))
        return 0;
    const k = toKey(stcks, hall);
    const m = memo.get(k);
    if(m != undefined)
        return m;
    let ans = 1000*1000*1000;

    for(let st = 0; st < stcks.length; st++) {
        if(stcks[st].length == 0)
            continue;
        if(stackPseudoSolved(stcks, st))
            continue;
        const v = stcks[st][stcks[st].length - 1];
        const hidx = hallIdx(st);
        const hallPoss = movesToHall(hall, hidx);
        for(const h of hallPoss) {
            hall[h] = v
            stcks[st].pop();

            const subCst = solve(stcks, hall, stackh);

            stcks[st].push(v);
            hall[h] = '.';

            const cst = subCst + costMp[v] * (1 + stackh - stcks[st].length + Math.abs(h - hidx));
            ans = cst < ans ? cst : ans;
        }
    }

    for(let i = 0; i < hall.length; i++) {
        const h = hall[i];
        if(h == '.')
            continue;
        const trgt = h.charCodeAt(0) - 'A'.charCodeAt(0);
        const hidx = hallIdx(trgt);
        if(stcks[trgt].length == stackh)
            continue;
        if(!stackPseudoSolved(stcks, trgt))
            continue;
        const slots = openHallSlots(hall, i);
        if(!slots.includes(hidx))
            continue;

        hall[i] = '.'
        stcks[trgt].push(h);
        const subCst = solve(stcks, hall, stackh);
        hall[i] = h
        stcks[trgt].pop();

        const cst = subCst + costMp[h] * (Math.abs(i - hidx) + stackh - stcks[trgt].length);
        ans = cst < ans ? cst : ans;
    }

    memo.set(k, ans);
    return ans;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

async function main() {
    const stcks: elt[][] = [[], [], [], []];

    for await (const line of rl) {
        const stck: elt[] = [];
        const mtches = Array.from(line.matchAll(/[A-Z]/g));
        for(const i in mtches)
            stcks[i].push(mtches[i][0] as elt);
    }
    for(const i in stcks)
        stcks[i].reverse();
    const hall = Array(11).fill('.');
    console.log(`Part 1: ${solve(stcks, hall, 2)}`);

    const stckD: elt[][] = [['D', 'D'], ['B', 'C'], ['A', 'B'], ['C', 'A']];
    for(const i in stcks)
        stcks[i] = [stcks[i][0]].concat(stckD[i]).concat([stcks[i][1]]);
    console.log(`Part 2: ${solve(stcks, hall, 4)}`);
}

main();

File: AdventOfCode2021/pysolns/d23.py
import sys
from re import findall

INF = 10**9

def lines():
  return [line.strip() for line in sys.stdin]

l = [findall(r'[A-Z]', l) for l in lines()]
l1, l2 = l[2], l[3]
stcks = [[l2[x], l1[x]] for x in range(4)]

def hallIdx(stck):
  return 2 * (1 + stck)

def isRoomDoor(idx):
  return 0 < idx < 10 and idx % 2 == 0

def openHallSlots(hall, idx):
  s = []
  w = idx
  while w > 0 and hall[w-1] == '.':
    w -= 1
    s.append(w)
  w = idx
  while w < 10 and hall[w+1] == '.':
    w += 1
    s.append(w)
  return s

def movesToHall(hall, idx):
  return [pos for pos in openHallSlots(hall, idx) if not isRoomDoor(pos)]

def stackPseudoSolved(stcks, st):
  tgt = chr(ord('A') + st)
  return all(x == tgt for x in stcks[st])

def stackSolved(stcks, st, stackh):
  return len(stcks[st]) == stackh and stackPseudoSolved(stcks, st)

def solved(stcks, hall, stackh):
  s = all(stackSolved(stcks, st, stackh) for st in range(4))
  return s

def toKey(stcks, hall):
  return tuple([tuple([tuple(st) for st in stcks]), tuple(hall)])

costMp = {
    'A': 1,
    'B': 10,
    'C': 100,
    'D': 1000,
}

memo = {}
def tryMoves(stcks, hall, stackh):
  if solved(stcks, hall, stackh):
    return 0
  k = toKey(stcks, hall)

  if k in memo:
    return memo[k]

  ans = INF
  # Move from stacks to hall
  for st in range(4):
    if len(stcks[st]) == 0:
      continue
    if stackPseudoSolved(stcks, st):
      continue
    v = stcks[st][-1]
    hidx = hallIdx(st)
    hallPoss = movesToHall(hall, hidx)
    for h in hallPoss:
      hall[h] = v
      stcks[st].pop()

      subcst = tryMoves(stcks, hall, stackh)

      stcks[st].append(v)
      hall[h] = '.'

      mvcst = costMp[v] * (1 + stackh - len(stcks[st]) + abs(h - hidx))
      ans = min(ans, mvcst + subcst)

  # Move from hall to stacks
  for i, h in enumerate(hall):
    if h == '.':
      continue
    tgt = ord(h) - ord('A')
    hidx = hallIdx(tgt)
    if len(stcks[tgt]) == stackh:
      continue
    if not stackPseudoSolved(stcks, tgt):
      continue
    reachable = openHallSlots(hall, i)
    if hidx not in reachable:
      continue

    hall[i] = '.'
    stcks[tgt].append(h)
    subcst = tryMoves(stcks, hall, stackh)
    hall[i] = h
    stcks[tgt].pop()

    mvcst = costMp[h] * (abs(i - hidx) + stackh-len(stcks[tgt]))
    ans = min(ans, subcst + mvcst)

  memo[k] = ans
  return ans

def partA():
  print('Part 1:', tryMoves(stcks, ['.']*11, 2))

def partB():
  ls = ['DD', 'BC', 'AB', 'CA']
  for i in range(4):
    stcks[i] = [stcks[i][0]] + list(ls[i]) + [stcks[i][1]]
  print('Part 2:', tryMoves(stcks, ['.']*11, 4))

partA()
partB()
